tools/document-pii/src/pii_finder.rs

```diff
@@ -0,0 +1,266 @@
+                    ident.push_str("::");
+                    ident.push_str(&second_segment.ident.to_string());
+                    segments.push(ident);
+                } else {
+                    segments.push(ident);
+                }
+            }
+        }
+        _ => {
+            use quote::ToTokens;
+            let mut tokens = proc_macro2::TokenStream::new();
+            ty.to_tokens(&mut tokens);
+            segments.push(tokens.to_string());
+        }
+    }
+}
+
+/// Checks if an attribute is equal to a specified name and value.
+fn has_attr_value(attr: &Attribute, name: &str, value: &str) -> bool {
+    if let Meta::List(meta_list) = &attr.meta {
+        if meta_list.path.is_ident("metastructure") {
+            let mut iter = meta_list.tokens.clone().into_iter();
+            while let Some(token) = iter.next() {
+                if let proc_macro2::TokenTree::Ident(ident) = &token {
+                    if ident == name {
+                        if let Some(proc_macro2::TokenTree::Punct(punct)) = iter.next() {
+                            if punct.as_char() == '=' {
+                                if let Some(proc_macro2::TokenTree::Literal(lit_val)) = iter.next()
+                                {
+                                    let mut lit_val = lit_val.to_string();
+                                    lit_val.pop(); // remove superfluous quotes. "\"true\"" -> "true"
+                                    lit_val.remove(0);
+                                    if lit_val == value {
+                                        return true;
+                                    }
+                                }
+                            }
+                        }
+                    };
+                }
+            }
+        }
+    }
+    false
+}
+
+/// Checks if a field has the metastructure "pii" and if it does, if it is equal to any
+/// of the values that the user defines.
+fn has_pii_value(pii_values: &Vec<String>, field: &Field) -> bool {
+    for attr in &field.attrs {
+        for pii_value in pii_values {
+            if has_attr_value(attr, "pii", pii_value) {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+/// Finds all pii-fields of either a single type if provided, or of all the defined types in the
+/// rust project.
+pub fn find_pii_fields(
+    item: Option<&str>,
+    all_types: &HashMap<String, EnumOrStruct>,
+    use_statements: &BTreeMap<String, BTreeSet<String>>,
+    pii_values: &Vec<String>,
+) -> anyhow::Result<BTreeSet<Vec<TypeAndField>>> {
+    let mut pii_types = BTreeSet::new();
+    let mut current_path = vec![];
+    match item {
+        Some(path) => {
+            if let Some(structorenum) = { all_types.get(path).cloned() } {
+                let module_path = path
+                    .rsplit_once("::")
+                    .ok_or_else(|| anyhow!("invalid module path: {}", path))?
```

Comment:
I'm not familiar with the `anyhow` usage, but is this how we're supposed to use it? My understanding is that you let the error pop up and add more context with `.context()`.

Reply:
do you mean like appending multiple error messages instead of exiting at the first error ?